core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +240,25 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if one side is smaller than the other by more than SIZE_DIFFERENCE_THRESHOLD times.
+        // We use 8x factor because the reduction in size from aggregation and filtering
```

Comment:
> We use 8x factor because the reduction in size from aggregation and filtering

Could you elaborate a bit more? If aggregation or filtering is properly estimated, then 8x factor shouldn't be needed

Reply:
We will reach the size based estimation logic only in the presence of some unestimated term. Since we don't know the full impact of that unestimated term, I'm being conservative by changing plan only if the size difference is big.

Reply:
> Since we don't know the full impact of that unestimated term,

but unestimated term can be reducing only (in terms of rows) and we use `getFirstKnownOutputSizeInBytes` for estimation (so estimated terms)

Reply:
Unlike the broadcast join logic we're not just checking that the output size is below some static threshold here. We're comparing two estimates both of which have potentially unaccounted terms. We don't know whether one side's reduction could have been greater than the other's if all the estimates were present (e.g. filters of very different selectivities).